### PR TITLE
Added multi-lane support to BasecallsConverter

### PR DESCRIPTION
**Pull request by jacarey**
### Description

This PR allows for the specification of multiple lanes when using IlluminaBasecallsToFastq or IlluminaBasecallsToSam. 

Reads with the same sample barcodes that span multiple lanes are output to the same files.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

**Comment by jacarey** [comment]
Added Lane 2 test files which are a copy of the Lane 1 data.

**Comment by kbergin** [comment]
Apologies for the delay, Jay - feel free to tag me whenever you make these PRs related to the basecalling efforts. George is out today but I'll make sure he's aware of it tomorrow so we can get this reviewed.

File: src/main/java/picard/illumina/ClusterDataToSamConverter.java [hunk 1]
```diff
@@ -104,21 +104,21 @@ public String getHelpDoc() {
      * Constructor
      *
      * @param runBarcode                Used to construct read names.
-     * @param readGroupId               If non-null, set RG attribute on SAMRecord to this.
+     * @param laneToReadGroup               If non-null, set RG attribute on SAMRecord to this.
```
Comment (gbggrant) [inline src/main/java/picard/illumina/ClusterDataToSamConverter.java:107]: extra tab here.

File: src/main/java/picard/illumina/IlluminaBasecallsToFastq.java [hunk 2]
```diff
@@ -111,7 +111,7 @@
     public File BARCODES_DIR;
 
     @Argument(doc = "Lane number. ", shortName = StandardOptionDefinitions.LANE_SHORT_NAME)
```
Comment (gbggrant) [inline src/main/java/picard/illumina/IlluminaBasecallsToFastq.java:113]: Maybe update usage here to indicate that more than one lane may be provided?

File: src/main/java/picard/illumina/IlluminaBasecallsToSam.java [hunk 3]
```diff
@@ -122,7 +123,7 @@
     public File BARCODES_DIR;
 
     @Argument(doc = "Lane number. ", shortName = StandardOptionDefinitions.LANE_SHORT_NAME)
```
Comment (gbggrant) [inline src/main/java/picard/illumina/IlluminaBasecallsToSam.java:125]: Maybe update usage here to indicate that more than one lane may be provided?

File: src/test/java/picard/illumina/IlluminaBasecallsToSamTest.java [hunk 4]
```diff
@@ -176,29 +176,30 @@ public void testMolecularBarcodes(final String readStructure, final String[] umi
...
-                {"singleBarcodeAltName.", "multiplexed_positive_rgtags.params", 1, "25T8B25T", BASECALLS_DIR, TEST_DATA_DIR, null},
-                {"dualBarcode.", "library_double.params", 2, "25T8B8B25T", DUAL_BASECALLS_DIR, DUAL_TEST_DATA_DIR, null},
-                {"cbclConvert.", "library_double.params", 2, "151T8B8B151T", TEST_DATA_DIR_WITH_CBCLS, DUAL_CBCL_TEST_DATA_DIR, null},
-                {"hiseqxSingleLocs.", "library_double.params", 2, "25T8B8B25T", TEST_DATA_HISEQX_SINGLE_LOCS, HISEQX_TEST_DATA_DIR, null},
-                {"hiseqxSingleLocs.", "library_double.params", 2, "25T8B8B25T", TEST_DATA_HISEQX_SINGLE_LOCS, HISEQX_TEST_DATA_DIR, null},
-                {"dualBarcode.", "library_double.params", 2, "25T8B8B25T", DUAL_BASECALLS_DIR, DUAL_TEST_DATA_DIR, 1101},
-                {"cbclConvert.", "library_double.params", 2, "151T8B8B151T", TEST_DATA_DIR_WITH_CBCLS, DUAL_CBCL_TEST_DATA_DIR, 1102}
+         /*      {"multiplexedBarcode.", "library.params", 1, "25T8B25T", BASECALLS_DIR, new File(TEST_DATA_DIR.getParentFile(),"sams_with_DS"), null, new int[]{1}},
```
Comment (gbggrant) [inline src/test/java/picard/illumina/IlluminaBasecallsToSamTest.java:185]: Is this commented out for a reason?  If it is no longer needed in the test, can you delete it / explain?

**Review by gbggrant — APPROVED** [review]
Looks good.  But there are a lot of empty files added for testing.  Would it be possible to just create them during the test?  Maybe build the lane structure you need in a temp directory and then copy over the non-empty versions and create the empties?  Just a thought.

**Comment by jacarey** [comment]
> Looks good. But there are a lot of empty files added for testing. Would it be possible to just create them during the test? Maybe build the lane structure you need in a temp directory and then copy over the non-empty versions and create the empties? Just a thought.

I got rid of all the empty files and changed the tests so we don't bother checking empty fastqs at all.

**Review by kachulis — COMMENTED** [review]
A few minor nits.   Also, I wonder if it would be a little cleaner to push the multi-lane capability down into the IlluminaDataProviderFactory/IlluminaDataProvider classes, so that the eventual looping over lanes lives there, instead of in the BasecallConverter implementations.   There are a couple of spots with things like `this.tiles = laneFactories[0].getAvailableTiles();` that might be a little cleaner if done that way instead.

**Comment by jacarey** [comment]
> A few minor nits. Also, I wonder if it would be a little cleaner to push the multi-lane capability down into the IlluminaDataProviderFactory/IlluminaDataProvider classes, so that the eventual looping over lanes lives there, instead of in the BasecallConverter implementations. There are a couple of spots with things like `this.tiles = laneFactories[0].getAvailableTiles();` that might be a little cleaner if done that way instead.

I would have liked to push it down in the data provider but it currently has file mappings in data structures that would collide with multiple lanes. It would require a pretty big refactor to do that.

**Comment by gbggrant** [comment]
@jacarey looks good - if you want to update with the base we can go ahead and merge.  Unclear to me as to what the order of PRs you want to merge.

**Comment by jacarey** [comment]
@gbggrant There is going to be a bit of ugly merging here. I think we want to merge this one first.

**Comment by jacarey** [comment]
@gbggrant I think we are all set with this one too. Then I'll rebase into the final big one.

**Comment by gbggrant** [comment]
@jacarey - okay to merge this?

**Comment by jacarey** [comment]
@gbggrant good to go

**Pull request merged**